tests/doc_test.go

```diff
@@ -0,0 +1,246 @@
+					fmt.Errorf("%s without a preceeding %s", bt, ZQ_COMMAND)
+			}
+			examples = append(examples, ZQExamplePair{command, fcb})
+			command = nil
+			// A fenced code block need not specify an info string, or it
+			// could be arbitrary. The default case is to ignore everything
+			// else.
+		}
+		return ast.WalkContinue, nil
+	})
+
+	if command != nil && err == nil {
+		err = fmt.Errorf("%s without a following %s", ZQ_COMMAND, ZQ_OUTPUT)
+	}
+	return examples, err
+}
+
+// Return the text of a ast.FencedCodeBlock as a string.
+func BlockString(fcb *ast.FencedCodeBlock, source []byte) string {
+	var b strings.Builder
+	for i := 0; i < fcb.Lines().Len(); i++ {
+		line := fcb.Lines().At(i)
+		b.Write(line.Value(source))
+	}
+	return b.String()
+}
+
+// Remove doc author decorations of ZQ output in markdown text so it can be
+// compared with actual ZQ output.
+func TrimBlockDecorations(output string) string {
+	return strings.TrimSuffix(output, "...\n")
+}
+
+// Generate ZQ example test cases from ZQ example pairs found in markdown.
+func TestcasesFromFile(filename string) ([]ZQExampleTest, error) {
+	var tests []ZQExampleTest
+	var examples []ZQExamplePair
+	var b strings.Builder
+	source, err := ioutil.ReadFile(filename)
+	if err != nil {
+		return nil, err
+	}
+	reader := text.NewReader(source)
+	parser := goldmark.DefaultParser()
+	doc := parser.Parse(reader)
+	examples, err = CollectExamples(doc, source)
+	if err != nil {
+		return nil, err
+	}
+	for i, example := range examples {
+		b.Reset()
+		b.WriteString(filename)
+		b.WriteString("-")
+		b.WriteString(strconv.Itoa(i + 1))
+		name := b.String()
+		command := strings.TrimSpace(BlockString(example.command, source))
+		// Even though we're using TrimBlockDecorations(), we
+		// don't support head.
+		output := TrimBlockDecorations(BlockString(example.output, source))
+		tests = append(tests, ZQExampleTest{name, command, output})
+	}
+	return tests, nil
+}
+
+func MarkdownFiles() []string {
+	// This needs to find markdown files in the repo. Right now we just
+	// find one.
+	return []string{
+		"../zql/docs/processors/README.md",
+		"../zql/docs/search-syntax/README.md",
+	}
+}
+
+func TestMarkdownExamples(t *testing.T) {
+	pwd, err := os.Getwd()
+	require.Equal(t, nil, err)
+	defer os.Chdir(pwd)
+	zq, err := filepath.Abs("../dist/zq")
+	require.Equal(t, nil, err)
+	var alltests []ZQExampleTest
+	for _, filename := range MarkdownFiles() {
+		tests, err := TestcasesFromFile(filename)
+		require.Equal(t, nil, err, "error when processing file %s: %v", filename, err)
+		alltests = append(alltests, tests...)
+	}
+	err = os.Chdir(SAMPLEDATA_PATH)
+	require.Equal(t, nil, err)
+	for _, test := range alltests {
+		// TODO:
+		// 1. This stuff should be folded into pkg/test somehow. Let's
+		// wait on review comments on the other stuff first.
+		// 2. Sanitize command.
+		command := strings.ReplaceAll(test.command, "zq ", zq+" ")
+		c := exec.Command("bash", "-c", command)
```

Comment:
The bash usage is for globbing support in the commands, is that right? I see you have the todo for sanitizing this. My suggestion is that we initially use only path.filepath.Glob ( https://golang.org/pkg/path/filepath/#Glob ) & exec `zq` directly, even if that restricts the types of examples we can use for now.

Reply:
Not a push back but instead some explanation on reasoning for using bash -c:

In addition to file globbing, the other reason I was using bash was forward thinking: it was to support many examples' having only a small portion of their returning record. For example see thing one: https://github.com/mccanne/zq/tree/a62cf84c418a72e8cf66fbcbd20340b74683083c/zql/docs/processors#example

The example is ended with dots (for ellipsis) to indicate a continuation of the records.

@philrz has a desire for the doc examples to have as little as zql as needed, so he doesn't want his examples to have head procs except when demonstrating the head proc. My idea was to support an additional optional annotation in the info strings that support an effective "head -n N" via bash.

An example like:
~~~
```zql-command head:5
zq -f table '* | cut ts,uid' conn.log
```
~~~
Would run: `bash -c "zq -f table '* | cut ts,uid' conn.log | head -n 5"`

It is a convenient way to get what I want without having to do a lot of post processing of rows and allow the doc author to have a clean example.

Reply:
Thanks for capturing this so well @mikesbrown. :)

Reply:
understood, and I agree that we shouldn't have the `| head 10` on all the examples. Let's do the output truncation in the testing support code. It will be less anxiety inducing that needing to sanitize the arguments to bash to prevent abuse.

Reply:
Done